### PR TITLE
Fixes #13417: Provide facet registration through plugin API

### PR DESCRIPTION
**Pull request by ehelms**
This sprang out of this discussion -- https://github.com/Katello/katello/pull/5724#discussion_r50908625

### CODE REVIEW

**Comment by domcleal** [comment]
Please check the test failures.

**Comment by ehelms** [comment]
@ShimShtein is there a way to unregister or cleanly delete a facet? If you look at my test file, the other tests fail because they find my fake Facet I needed to create. My attempts to remove it have thus far failed.

**Comment by ShimShtein** [comment]
@ehelms Unfortunately there isn't. You can take a look [here](https://github.com/theforeman/foreman/blob/develop/test/unit/facet_test.rb#L17) to see what leftovers you have to remove by hand. I am not sure it's worth creating an "unregister" method, since the use case is valid only for framework testing. Any other facet should remain registered through the whole lifetime of the tests.

**Comment by ShimShtein** [comment]
Basically I don't have an objection to add it to the plugin interface as long as it is available for the core too. The reason for the non-plugin interface is for core features that are implemented as facets. Right now it would be puppet, until it moves out of core to a plugin. Compute resources are a good candidate for getting out to facets too.
So it would be :+1: for the PR.

**Comment by ehelms** [comment]
Updated the test that was failing.

File: test/unit/plugin_test.rb [hunk 1]
```diff
@@ -279,4 +280,19 @@ def test_extend_page
     assert_equal 1, ::Pagelets::Manager.sorted_pagelets_at("tests/show", :main_tabs).count
     assert_equal "My Tab", ::Pagelets::Manager.sorted_pagelets_at("tests/show", :main_tabs).first.name
   end
+
+  def test_register_facet
+    Facets.stubs(:configuration).returns({}
```
Comment (domcleal) [inline test/unit/plugin_test.rb]: missing `)`

**Comment by domcleal** [comment]
There are a rather large number of test failures still.  I'm not sure if it's caused by the syntax error or something else though.

**Comment by domcleal** [comment]
Merged as 7d90b5c, thanks @ehelms.

Could you please update the docs at http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin?

**Pull request closed without merging**